File: sdk/ts/src/modules/access.ts
import { getContract, type Address } from 'viem';
import { AgentAccessAbi } from '../abis/AgentAccess.js';
import { sendWrite, type Context } from '../internal.js';
import type { TxResult } from '../types.js';

/** Largest uint48 — an authorization that never expires. */
export const NO_EXPIRY = 281_474_976_710_655;

export interface AccessModule {
  /** Principal authorizes a hot operator key until `expiry` (unix seconds, default {@link NO_EXPIRY}). */
  authorizeOperator(operator: Address, expiry?: number): Promise<TxResult>;
  revokeOperator(operator: Address): Promise<TxResult>;
  /**
   * Signed by the operator itself: give up its own authorization for `agent`. Lets a hot key that
   * may have leaked cut itself off without waiting for the principal to act.
   */
  renounceOperator(agent: Address): Promise<TxResult>;
  /** True when `caller` is `agent` itself or one of its unexpired operators. */
  isOperatorFor(agent: Address, caller: Address): Promise<boolean>;
  /**
   * Unix second at which `operator`'s authorization for `agent` lapses, or 0 when there is no
   * live authorization — never granted, revoked, renounced, or simply expired. A non-zero value
   * is therefore always in the future, so this and {@link AccessModule.isOperatorFor} agree.
   */
  operatorExpiry(agent: Address, operator: Address): Promise<number>;
}

export function createAccessModule(ctx: Context): AccessModule {
  const address = ctx.addresses.access;
  const reader = getContract({ address, abi: AgentAccessAbi, client: ctx.publicClient });

  return {
    async authorizeOperator(operator, expiry = NO_EXPIRY) {
      return sendWrite(ctx, address, AgentAccessAbi, 'authorizeOperator', [operator, expiry]);
    },
    async revokeOperator(operator) {
      return sendWrite(ctx, address, AgentAccessAbi, 'revokeOperator', [operator]);
    },
    async renounceOperator(agent) {
      return sendWrite(ctx, address, AgentAccessAbi, 'renounceOperator', [agent]);
    },
    isOperatorFor: (agent, caller) => reader.read.isOperatorFor([agent, caller]),
    operatorExpiry: (agent, operator) => reader.read.operatorExpiry([agent, operator]),
  };
}

File: sdk/python/nexusweb3/contracts/access.py
"""AgentAccess — the operator registry (`IAgentAccess`)."""

from __future__ import annotations

from web3 import Web3

from ..tx import TxResult
from ..types import MAX_UINT48
from .base import ContractClient

__all__ = ["AccessClient"]


class AccessClient(ContractClient):
    """Principals authorize hot operator keys here; every other v2 contract reads this registry."""

    #: `type(uint48).max` — an authorization that never expires.
    NO_EXPIRY = MAX_UINT48

    def authorize_operator(self, operator: str, expiry: int = MAX_UINT48) -> TxResult:
        """Let `operator` act for the signing principal until `expiry` (unix seconds)."""
        return self._send("authorizeOperator", Web3.to_checksum_address(operator), int(expiry))

    def revoke_operator(self, operator: str) -> TxResult:
        return self._send("revokeOperator", Web3.to_checksum_address(operator))

    def renounce_operator(self, agent: str) -> TxResult:
        """Signed by the operator itself: give up its own authorization for `agent`.

        Lets a hot key that may have leaked cut itself off without waiting for the principal.
        """
        return self._send("renounceOperator", Web3.to_checksum_address(agent))

    def is_operator_for(self, agent: str, caller: str) -> bool:
        return bool(
            self._call("isOperatorFor", Web3.to_checksum_address(agent), Web3.to_checksum_address(caller))
        )

    def operator_expiry(self, agent: str, operator: str) -> int:
        """Unix second the authorization lapses, or 0 when there is no live one.

        0 covers every way an authorization can be absent: never granted, revoked, renounced, or
        simply expired. A non-zero result is therefore always in the future, so this and
        :meth:`is_operator_for` never disagree.
        """
        return int(
            self._call("operatorExpiry", Web3.to_checksum_address(agent), Web3.to_checksum_address(operator))
        )
